rfcs/0096-embedded-rust-interface-api.md

```diff
@@ -0,0 +1,333 @@
+
+  let mut workspace = ultimate_cat_spotter::Workspace::new(
+    &mut input_data,
+    &mut output_data,
+  );
+
+  let mut model = ultimate_cat_spotter::Model::new();
+  model.infer(&mut workspace, TVMDevice::CPU);
+  
+  assert_eq!(output_data, my_app_logic::expected_output());
+}
+```
+
+## Running a model with workspace pools
+This extends the above premise and provides an additional memory pool argument:
+
+```rust
+mod my_app_logic;
+extern crate tvmgen_ultimate_cat_spotter as ultimate_cat_spotter;
+
+fn main() {
+  let mut input_data: [i8; 25600] = my_app_logic::create_input();
+  let mut output_data: [f32; 12] = my_app_logic::create_output();
+  let mut memory_pool: [u8; 20000] = my_app_logic::create_memory_pool();
+
+  let mut workspace = ultimate_cat_spotter::Workspace::new(
+    &mut input_data,
+    &mut output_data,
+    &mut memory_pool,
+  );
+
+  let mut model = ultimate_cat_spotter::Model::new();
+  model.infer(&mut workspace, TVMDevice::CPU);
+  assert_eq!(output_data, my_app_logic::expected_output());
+}
+```
+
+## Running a model using constant pools
+By utilising macros, users can generate the appropriate constant pools at compilation time:
+
+```rust
+mod my_app_logic;
+extern crate tvmgen_ultimate_cat_spotter as ultimate_cat_spotter;
+
+fn main() {
+  let mut input_data: [i8; 25600] = my_app_logic::create_input();
+  let mut output_data: [f32; 12] = my_app_logic::create_output();
+  let mut memory_pool: [u8; 20000] = my_app_logic::create_memory_pool(tvmgen_default::constant_pool_data!());
+
+  let mut workspace = ultimate_cat_spotter::Workspace::new(
+    &mut input_data,
+    &mut output_data,
+    &mut memory_pool,
+  );
+
+  let mut model = ultimate_cat_spotter::Model::new();
+  model.infer(&mut workspace, TVMDevice::CPU);
+  assert_eq!(output_data, my_app_logic::expected_output());
+}
+```
+
+## Running a model with I/O memory pools
+This utilises Rust slices to take an input array of bytes and provide Rust access to different sections within the memory pools:
+
+```rust
+mod my_app_logic;
+extern crate tvmgen_ultimate_cat_spotter as ultimate_cat_spotter;
+
+fn main() {
+  let mut memory_pool: [u8; 20000] = my_app_logic::create_memory_pool();
+
+  let mut workspace = ultimate_cat_spotter::Workspace::new(
+    &mut memory_pool,
+  );
+
+  let mut model_input = workspace.input_data();
+  my_app_logic::copy_input_data(model_input);
+
+  let mut model = ultimate_cat_spotter::Model::new();
+  model.infer(&mut workspace, TVMDevice::CPU);
+  assert_eq!(workspace.output_data(), my_app_logic::expected_output());
+}
+```
+
+## Rust Device API
+In the Rust interface, we provide a similar interface as the C Device API, providing a trait for driver authors to implement:
+
+```rust
+trait TVMDevice {
+    fn activate();
+    fn open();
+    fn close();
+    fn deactivate();
+}
+```
+
+This can then be used by a driver author as simply:
+
+```rust
+impl TVMDevice for MyDriver { ... }
+```
```

Comment:
Thakns for the RFC, after reading the proposed code, I think the main thing that would be worth clarifying is the overall scope, and scope limitations in relation with existing namespace. I will put my input in a followup comment section.